### ISSUE TITLE
Decklink Display inconsistent FPS

### ISSUE THREAD

**Issue by alatteri** [CESNET/UltraGrid#88]
Hello,

I am noticing that the Decklink Display FPS seem to be inconsistent with the Decklink Capture FPS (input signal).
I have machine A (encoder) capturing a 1920x1080 RGB10bit444 SDI 3G signal with x265.  The capture rate is pretty much always shown as 23.97.
On machine B (decoder) which is on the same network, output via Decklink the FPS display varies slightly, but enough that it shows visible stutter, as if some frames are being held an extra frame, or skipped. 

CPU does not seem to be the constraint on the encoder or decoder.  Outputting via SDL does not seem to have this issue as long as the monitor is running at matching HZ to the FPS.

Any thoughts?

Capture:
```
[2020-09-09 16:15:01] [Decklink capture] 120 frames in 5.00475 seconds = 23.9772 FPS
[2020-09-09 16:15:06] [Decklink capture] 120 frames in 5.00482 seconds = 23.9769 FPS
[2020-09-09 16:15:11] [Decklink capture] 120 frames in 5.00483 seconds = 23.9768 FPS
[2020-09-09 16:15:16] [Decklink capture] 120 frames in 5.00486 seconds = 23.9767 FPS
[2020-09-09 16:15:21] [Decklink capture] 120 frames in 5.00485 seconds = 23.9767 FPS
[2020-09-09 16:15:26] [Decklink capture] 120 frames in 5.00489 seconds = 23.9765 FPS
[2020-09-09 16:15:31] [Decklink capture] 120 frames in 5.00495 seconds = 23.9763 FPS
[2020-09-09 16:15:36] [Decklink capture] 120 frames in 5.00495 seconds = 23.9762 FPS
[2020-09-09 16:15:41] [Decklink capture] 120 frames in 5.00499 seconds = 23.9761 FPS
```

Display:
```
20-09-09 16:14:44] [Decklink display] 120 frames in 5.00865 seconds = 23.9586 FPS
[2020-09-09 16:14:49] [Decklink display] 121 frames in 5.03791 seconds = 24.0179 FPS
[2020-09-09 16:14:54] [Decklink display] 120 frames in 5.00609 seconds = 23.9708 FPS
[2020-09-09 16:14:59] [Decklink display] 120 frames in 5.00809 seconds = 23.9612 FPS
[2020-09-09 16:15:04] [Decklink display] 120 frames in 5.00108 seconds = 23.9948 FPS
[2020-09-09 16:15:09] [Decklink display] 120 frames in 5.00645 seconds = 23.9691 FPS
[2020-09-09 16:15:14] [Decklink display] 121 frames in 5.04708 seconds = 23.9742 FPS
[2020-09-09 16:15:19] [Decklink display] 120 frames in 5.00152 seconds = 23.9927 FPS
[2020-09-09 16:15:24] [Decklink display] 121 frames in 5.04194 seconds = 23.9987 FPS
[2020-09-09 16:15:29] [Decklink display] 120 frames in 5.01074 seconds = 23.9485 FPS
[2020-09-09 16:15:34] [Decklink display] 120 frames in 5.00553 seconds = 23.9735 FPS
[2020-09-09 16:15:39] [Decklink display] 120 frames in 5.00648 seconds = 23.969 FPS
[2020-09-09 16:15:44] [Decklink display] 120 frames in 5.00522 seconds = 23.975 FPS
[2020-09-09 16:15:49] [Decklink display] 120 frames in 5.00461 seconds = 23.9779 FPS
[2020-09-09 16:15:54] [Decklink display] 120 frames in 5.00547 seconds = 23.9738 FPS
[2020-09-09 16:15:59] [Decklink display] 120 frames in 5.00554 seconds = 23.9734 FPS
[2020-09-09 16:16:04] [Decklink display] 120 frames in 5.01256 seconds = 23.9399 FPS
[2020-09-09 16:16:09] [Decklink display] 121 frames in 5.03924 seconds = 24.0115 FPS
[2020-09-09 16:16:14] [Decklink display] 120 frames in 5.0069 seconds = 23.9669 FPS
```

**Comment by alatteri** [comment]
PS: I don't recall this issue prior to BMD SDK 11.6, but I haven't rolled back to test, so I could be wrong.

**Comment by alatteri** [comment]
Tested with H.264, similar results.

```
[2020-09-09 16:35:53] [Decklink display] 120 frames in 5.00676 seconds = 23.9676 FPS
[2020-09-09 16:35:58] [Decklink display] 120 frames in 5.00131 seconds = 23.9937 FPS
[2020-09-09 16:36:03] [Decklink display] 120 frames in 5.00937 seconds = 23.9551 FPS
[2020-09-09 16:36:08] [Decklink display] 120 frames in 5.00172 seconds = 23.9917 FPS
[2020-09-09 16:36:13] [Decklink display] 120 frames in 5.00569 seconds = 23.9727 FPS
[2020-09-09 16:36:18] [Decklink display] 120 frames in 5.0062 seconds = 23.9703 FPS
[2020-09-09 16:36:23] [Decklink display] 120 frames in 5.00472 seconds = 23.9774 FPS
[2020-09-09 16:36:28] [Decklink display] 120 frames in 5.00343 seconds = 23.9835 FPS
[2020-09-09 16:36:33] [Decklink display] 120 frames in 5.00515 seconds = 23.9753 FPS
[2020-09-09 16:36:38] [Decklink display] 120 frames in 5.00464 seconds = 23.9777 FPS
[2020-09-09 16:36:43] [Decklink display] 120 frames in 5.00498 seconds = 23.9761 FPS
[2020-09-09 16:36:48] [Decklink display] 120 frames in 5.00637 seconds = 23.9695 FPS
```

**Comment by MartinPulec** [comment]
The variance on receiver is quite natural, I wouldn't bother much - the rate is almost the same (the receiver has very very slightly lower, by 0.005%, but the sample is small). However, the stutter shouldn't occur.

Would it be possible try to use option "`no-low-latency`" for the DeckLink display and see if it doesn't help?

Also a signal that would show this may help - I've tried with `-t testcard:1920:1080:23.976:UYVY` but even the output looked quite jerky (but it may be given by the display).

**Comment by MartinPulec** [comment]
Just returning to that since it is still open:

> [...] that it shows visible stutter, as if some frames are being held an extra frame, or skipped.

I've thought about that and it is possible in the end. I've looked to the documentation and the synchronnous output seems to work in a way that it just pushes the frame to be displayed as next frame. A consequence of this, however, is that it perhaps replaces any frame that was planned before. Therefore it is a bit suspectible to timing jitter which may there occur in case of inter-frame compression.

That just for explanation. But as I responded in the original post, using `no-low-latency` option may help but I am not completly sure since it is not so much maintained since it is not the default value.

**Comment by TheSashmo** [comment]
I’ve been following this thread and I can say that I see a similar issue.  I even opened a ticket about it before, I just can’t recall which one it was.

The input and output logs showed the same frame rate identical to Alterris logs.  But I could visually see a stutter in the video.

I was not testing the same video format, but the input source was detected as 10bit 422.

The only way that I could avoid the frame drops was to change the override the input to 8bit 420.

If I had to guess, I suspect that ultragrid might  have an issue with 10bit 422.  The funny part was it doesn’t happen all the time.  I can test it in my lab and it can work just fine.  Once I connect it to some other real source like a production switcher the issue comes back.  I thought it might have been the production switcher but it was so random that I gave up and said I will live with 8bit 420.

I have to say that I think there is a deeper issue but I am in no position to pin point exactly where.  I can help provide remote setup and testing if needed.  But I don’t have access to any other 10bit 422 sources other than from a BM device, so hard to replicate what I was seeing.
